### PR TITLE
refactor(modal): rewrite with hook and support strict mode

### PR DESCRIPTION
**Pull request by hengkx**
### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / document update
- [ ] Component style update
- [ ] TypeScript definition update
- [x] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |  - Refactor Modal with hooks and support strict mode. <br />- Fix Modal `confirmLoading` button not reset.    |
| 🇨🇳 Chinese |  - 使用 Hooks 重构 Modal，使其兼容于严格模式。<br />- 修复 Modal 关闭后 `confirmLoading` 按钮没有复位的问题。       |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️ 

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/modal/ActionButton.tsx [hunk 1]
```diff
@@ -11,59 +10,53 @@ export interface ActionButtonProps {
...
-  loading: ButtonProps['loading'];
-}
-
-export default class ActionButton extends React.Component<ActionButtonProps, ActionButtonState> {
-  timeoutId: number;
-
-  clicked: boolean;
+const ActionButton: React.FC<ActionButtonProps> = props => {
+  const clickedRef = React.useRef<boolean>(false);
+  const ref = React.createRef<any>();
```
Comment (zombieJ) [inline components/modal/ActionButton.tsx:15]: useRef?

**Review by saeedrahimi — APPROVED** [review]

File: components/modal/Modal.tsx [hunk 2]
```diff
@@ -119,101 +119,102 @@ export interface ModalLocale {
...
-        footer={footer === undefined ? defaultFooter : footer}
-        visible={visible}
-        mousePosition={mousePosition}
-        onClose={this.handleCancel}
-        closeIcon={closeIconToRender}
-      />
-    );
-  };
+Modal.useModal = useModal;
+Modal.destroyAll = () => {};
```
Comment (zombieJ) [inline components/modal/Modal.tsx:209]: 空函数……

Comment (afc163) [reply components/modal/Modal.tsx:209]: 用例居然没挂。。

Comment (hengkx) [reply components/modal/Modal.tsx:209]: 这个是在index里面声明了

Comment (hengkx) [reply components/modal/Modal.tsx:209]: index 里面 把他给覆盖掉了 ，这边直接有ts声明 是必填的

Comment (zombieJ) [reply components/modal/Modal.tsx:209]: Modal 里不用写， index 里 as 转一下类型添加即可。

**Comment by afc163** [comment]
close #24382 too

**Pull request merged**